x/dex/keeper/grpc_query_get_twaps.go

```diff
@@ -0,0 +1,57 @@
+package keeper
+
+import (
+	"context"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/sei-protocol/sei-chain/x/dex/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (k Keeper) GetTwaps(goCtx context.Context, req *types.QueryGetTwapsRequest) (*types.QueryGetTwapsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	allRegisteredPairs := k.GetAllRegisteredPairs(ctx, req.ContractAddr)
+	twaps := []*types.Twap{}
+	for _, pair := range allRegisteredPairs {
+		prices := k.GetAllPrices(ctx, req.ContractAddr, pair)
+		twaps = append(twaps, &types.Twap{
+			Pair:            &pair,
+			Twap:            calculateTwap(ctx, prices, req.LookbackSeconds),
+			LookbackSeconds: req.LookbackSeconds,
+		})
+	}
+
+	return &types.QueryGetTwapsResponse{
+		Twaps: twaps,
+	}, nil
+}
+
+func calculateTwap(ctx sdk.Context, prices []*types.Price, lookback uint64) sdk.Dec {
+	// sort prices in descending order to start iteration from the latest
+	sort.Slice(prices, func(p1, p2 int) bool {
+		return prices[p1].SnapshotTimestampInSeconds > prices[p2].SnapshotTimestampInSeconds
+	})
+	var timeTraversed uint64 = 0
+	var weightedPriceSum sdk.Dec = sdk.ZeroDec()
+	for _, price := range prices {
+		newTimeTraversed := uint64(ctx.BlockTime().Unix()) - price.SnapshotTimestampInSeconds
+		if newTimeTraversed > lookback {
```

Comment:
in this case, wouldn't we exclude any data from the first time interval? 

Lets say for example we had data at T-61 and then T-59 ... T-0, when calculating the TWAP, it would technically only represent the weighted price sum for T-59 - T-0, right? because we would break on T-61 and never include data for the first interval?

Reply:
that's true. We should align on the behavior in such case between dex twap and oracle twap. Still factoring in T-61 but only assign a time weight of `1` sounds reasonable to me, if that's also how oracle twap is calculated

Reply:
yeah that makes sense to me, currently my implementation also has the missing first interval issue, but I can modify it to keep the first out of bounds data point for calculating to the start of the TWAP window

Reply:
will approve once this change is in

Reply:
pushed the fix as well as a unit test for this case